src/shared/logs-show.c

```diff
@@ -601,25 +616,36 @@ static int output_export(
                 sd_id128_to_string(boot_id, sid));
 
         JOURNAL_FOREACH_DATA_RETVAL(j, data, length, r) {
+                const char *c;
 
                 /* We already printed the boot id, from the data in
                  * the header, hence let's suppress it here */
                 if (length >= 9 &&
                     startswith(data, "_BOOT_ID="))
                         continue;
 
+                c = memchr(data, '=', length);
+                if (!c) {
+                        log_error("Invalid field.");
+                        return -EINVAL;
+                }
+
+                if (output_fields) {
+                        _cleanup_free_ char *n = NULL;
+
+                        n = strndup(data, c - (const char*) data);
+                        if (!n)
+                                return log_oom();
+
+                        if (!set_get(output_fields, n))
+                                continue;
+                }
```

Comment:
given that this code is used at various places in very similar ways, maybe introduce a new short helper function `field_set_test(Set *s, const char *s, size_t n)` that does this match check? Would also be a lot more readable...

Reply:
Not sure that we'd save that much. There's only two instances of this (in `output_json()` we already have a copy of the field name). Also, the test function could fail because of the `strndup` in there, turning the call site into something ugly like this:

```c
r = field_set_test(output_fields, data, eq - data);
if (r <0)
        return r;
if (r == 0)
        continue;
```

(Do you use `0` for successful tests or `1`?)

Let me know if you feel strongly about it, though, then I'll change it.

Reply:
I'd still prefer if this is moved into its own function. Besides being more readable (and shortening the already huge funciton a bit) I think it makes things a lot more readable. And also, it permits us to us strndupa() instead of strdup() (i.e. stack allocation instead of heap allocation — stack allocation using alloca()/strndupa() is not OK in loops), which is a really good idea, since we do this for every single field...